gbn: Sender side only packeting resend bug fix

ViktorTigerstrom opened the pull request:
Fixes #89 

This PR addresses the packeting resend bug through a sender only side fix.

This PR ensures that after we have resent the queue, we will wait until we know that both parties are in sync before we continue to send new packets. This ensures that we don't end up in an indefinitely resend loop due to latency and delayed NACKs by the other party, which could happen prior to this commit.

To understand why we need to await the awaited ACK/NACK after resending the queue, consider the following scenario:

1. Alice sends packets 1, 2, 3 & 4 to Bob.
2. Bob receives packets 1, 2, 3 & 4, and sends back the respective ACKs.
3. Alice receives ACKs for packets 1 & 2, but due to latency the ACKs for packets 3 & 4 are delayed and aren't received until Alice resend timeout has passed, which leads to Alice resending packets 3 & 4. Alice will after that receive the delayed ACKs for packets 3 & 4, but will consider that as the ACKs for the resent packets, and not the original packets which they were actually sent for. If we didn't wait after resending the queue, Alice would then proceed to send more packets (5 & 6).
4. When Bob receives the resent packets 3 & 4, Bob will respond with NACK 5. Due to latency, the packets 5 & 6 that Alice sent in step (3) above will then be received by Bob, and be processed as the correct response to the NACK 5. Bob will after that await packet 7.
5. Alice will receive the NACK 5, and now resend packets 5 & 6. But as Bob is now awaiting packet 7, this send will lead to a NACK 7. But due to latency, if Alice doesn't wait resending the queue, Alice will proceed to send new packet(s) before receiving the NACK 7.
6. This resend loop would continue indefinitely, so we need to ensure that Alice waits after she has resent the queue, to ensure that she doesn't proceed to send new packets before she is sure that both parties are in sync.

To ensure that we are in sync, after we have resent the queue, we will await that we EITHER:
1. Receive a NACK for the sequence number succeeding the last packet in the resent queue i.e. in step (3) above, that would be NACK 5.
2. Receive an ACK for the last packet in the resent queue i.e. in step (3) above, that would be ACK 4. After we receive the expected ACK, we will then wait for the duration of the resend timeout before continuing. The reason why we wait for the resend timeout before continuing, is that the ACKs we are getting after a resend, could be delayed ACKs for the
original packets we sent, and not ACKs for the resent packets. In step (3) above, the ACKs for packets 3 & 4 that Alice received were delayed ACKs for the original packets. If Alice would have immediately continued to send new packets (5 & 6) after receiving the ACK 4, she would have then received the NACK 5 from Bob which was the actual response to the resent queue. But as Alice had already continued to send packets 5 & 6 when receiving the NACK 5, the resend queue response to that NACK would cause the resend loop to continue indefinitely.

When either of the 2 conditions above are met, we will consider both parties to be in sync, and we can proceed to send new packets.

TODO:
- [X] Update the with extra docs + update commit messages with more detailed docs

ViktorTigerstrom commented on gbn/gbn_conn.go:
```diff
@@ -378,20 +425,179 @@ func (g *GoBackNConn) sendPacket(ctx context.Context, msg Message) error {
+		return nil
+	}
+
+	if g.sendQueue.isPayloadPackets(base, top) {
```
Due you think it makes sense to not require the resend catch up if we are sending only Ping packets @ellemouton? If you look at the current itest CI fail, there's some cases where this happens and we're resending the queue multiple times.

ViktorTigerstrom replied:
Thanks! Renamed this, and shifted the responsibility to the queue struct instead :)

ViktorTigerstrom replied:
Hmmm I'm not sure if this is commenting an old commit, but this should not be here any longer?

ellemouton replied:
it's still there: https://github.com/lightninglabs/lightning-node-connect/pull/87/files#diff-0d225f2c03c360d624ff3e8b46582178accd59eb318fb0ad8ef0e4f022abcd88R237-R247

ViktorTigerstrom replied:
Removed with the latest push, after testing that showed that this was unneccessary!

ViktorTigerstrom commented on gbn/gbn_conn.go:
```diff
@@ -507,6 +713,8 @@ func (g *GoBackNConn) receivePacketsForever() error { // nolint:gocyclo
 			g.pongTicker.Pause()
 		}
 
+		g.resendTicker.Reset(g.resendTimeout)
```
Do you agree that it makes sense to always reset the resend ticker if we receive any type of message? My reasoning behind this, is that if the other party actually sends something to us, we can be sure that they're actually still active. It might just be that the thread that's currently being prioritised for the other party currently, is the one that's resending the queue.

ellemouton replied:
yeah makes sense 👍

ViktorTigerstrom reviewed (commented):
Updated with docs, and made some small changes to the PR! This is now ready for review :fire:.

I'll address CI failures tomorrow, but the PR should be reviewable even before that happens.

I've confirmed locally through manual testing, that this PR causes no issues if the two parties have different resend timeouts. But, something I'd like to test is the interoperability if the client has updated to the new LNC version, but the server hasn't. Would be good to ensure that that doesn't cause more resends by the server than prior to this PR.

ellemouton commented on gbn/gbn_conn.go:
```diff
@@ -582,6 +788,13 @@ func (g *GoBackNConn) receivePacketsForever() error { // nolint:gocyclo
 			if gotValidACK {
 				g.resendTicker.Reset(g.resendTimeout)
 
+				g.awaitingCatchUpMu.RLock()
+				if m.Seq == g.awaitedACK && g.awaitingCatchUp {
+					log.Tracef("Got awaited ACK")
+					g.proceedAfterTime(g.catchUpId)
+				}
+				g.awaitingCatchUpMu.RUnlock()
```
that should also help with the removal of the need for the catch up ID

ViktorTigerstrom replied:
Commenting here regarding both this and (https://github.com/lightninglabs/lightning-node-connect/pull/87#discussion_r1389044721):

I've moved the logic into queue.ProcessACK instead!

1. Just to clarify, even though this does take an `RLock` before calling `proceedAfterTime`, it will not wait the entire timeout before releasing the `RLock`. This is since `time.AfterFunc` executes the callback in a separate `go-routine`, so the `RLock` will be released before the timeout has elapsed.

2. This doesn't remove the need for the `catchUpID`. Though I agree that maybe the `catchUpID` is overkill, and I can remove it if you think it just complicates more than it helps. I think I need to examplify with a scenario where the `catchUpID` is useful, to understand why I've added it. I should probably also add that I've experienced this while testing locally, so it does happen some times:
2a. Alice sends packets 1, 2 & 3 to Bob, but due latency she resend timeouts before receiving ACK 2 & 3 from Bob. Alice therefore resends packets 2 & 3.
2b. Alice now received the delayed ACKS 2 & 3 from Bob, and therefore executes `proceedAfterTime` for ACK 3 which starts the timeout before the it executes the callback. The callback will send the `awaitedACKSignal` once executed, if `awaitingCatchUp` is `true`.
2c. Alice now receives the NACK 4 that the resent packets in step (2a) led to. She now marks the resend catch up as done, and proceeds to send packets 4, 5 & 6.
2d. Due to packet loss, packet 5 never reaches Bob, so when Bob receives packet 6, he will respond with NACK 5.
2e. Alice now resends her queue again as the response to NACK 5, and then starts awaiting the catch up, and marks `awaitingCatchUp` as `true`.
2f. The callback from step `2b` now gets executed as the timeout has passed. As the the `awaitingCatchUp` is currently set to `true`, this callback will now mark the catch up incorrectly as done, and send the `awaitedACKSignal`. However, this callback that was executed was actually intended for the previous catch up cycle, and not the current cycle, so this callback shouldn't have marked the catch up as done.

This is what the `catchUpID` is used for. The current code instead also looks that the `catchUpID` is the still the same before marking the catch up as done, when a callback is eventually executed. Had that check existed for the above example, (2f) would not have marked the catch up as done.

Though I agree that it's maybe too complicated to understand why the `catchUpID` is needed. Its worth adding that the worst thing that likely can happen if we remove it, is that cases like the above would result in one extra resend cycle, so that's maybe not too bad compared to the complexity it adds.

Do you think I should still remove it or not, given the above explanation?

ellemouton replied:
> 2e. Alice now resends her queue again as the response to NACK 5, and then starts awaiting the catch up, and marks awaitingCatchUp as true.

Since Alice is currently busy with a resend/catch up, she should not re-trigger one here. 
I think the main thing leading to the added complexity is that the GBN is the one the the 'resendTimeout' and thus it looks like we are trying to let the queue work around this. How about instead letting the queue handle the when `resend` is allows to even be called/ or when the resent timeout ticks? 

I think a state machine approach would be best here. If we are in a resend/catch up state - then all parts should know this and act accordingly. 

Rather than adding complexity to fix this, I think we should first focus on reducing complexity 

ViktorTigerstrom replied:
> Rather than adding complexity to fix this, I think we should first focus on reducing complexity 

I definitely agree with that we should attempt to reduce  complexity, and not add more complexity. And the `CatchUpID` was clearly a concept that was really complicated. I've now removed the `CatchUpID` and replaced this with an `caughtUpSignal` channel instead.
This new channel marks that we've reached the caught up state, so that any past delayed go routine knows that the resend cycle it was initiated in has already been completed, and that the goroutine therefore should be a no-op. I hope this reduces the complexity and makes the code more understandable. I hope this matches what you'd want, but let me know if you have an even more clean solution :).

> Since Alice is currently busy with a resend/catch up, she should not re-trigger one here.

The problem here is that Alice actually isn't busy with the resend/catch up before she re-triggers the resend in (2e). She correctly marked the resend catch up as completed in (2c) when she got the expected NACK 4. 

The issue is that even though she did receive the NACK 4 and marked the resend catch up as completed, the delayed callback that was initiated with the `time.AfterFunc` execution in `proceedAfterTime` in (2b) has not been executed yet, and will still be executed after the resendTimeout has passed. This is why we need to ensure that this delayed callback won't mess anything up once it finally executes, if we've entered a new resend catch up cycle once the delayed callback is finally executed.

If you know a way that would make it possible to stop any delayed callbacks that were initiated during a resend catch up cycle from actually executing, if the resend cycle has been completed, that would be a much better solution :)! Else I think this new `caughtUpSignal` channel is relatively easy to understand.

ellemouton replied:
just make the delayed call back check that the "state" of the catch up is still the same as when it was executed?

ViktorTigerstrom replied:
This was now changed with the latest push to use a goroutine instead, so that the function is cancelled synchronously if it's still being executed when the resend catch up is over. Thanks for the help in simplifying this :)!

ellemouton commented on gbn/gbn_conn.go:
```diff
@@ -582,6 +788,13 @@ func (g *GoBackNConn) receivePacketsForever() error { // nolint:gocyclo
 			if gotValidACK {
 				g.resendTicker.Reset(g.resendTimeout)
 
+				g.awaitingCatchUpMu.RLock()
```
it's scary to me to hold a lock during a time.Sleep - if we have a subsystem handling these things rather, then it will make it easier to just swallow things during the wait

ViktorTigerstrom replied:
See https://github.com/lightninglabs/lightning-node-connect/pull/87#discussion_r1391462799 (1)

ellemouton reviewed (commented):
Nice investigation!!! 🎉 

Left a bunch of comments - basically, I think we can simplify and modularise the logic quite a bit by letting the `queue` handle everything.

ellemouton commented on gbn/gbn_conn.go:
```diff
@@ -378,20 +425,179 @@ func (g *GoBackNConn) sendPacket(ctx context.Context, msg Message) error {
+		return err
+	}
+
+	if !g.awaitingCatchUp {
```
this is causing a race condition

ViktorTigerstrom commented on gbn/gbn_conn.go:
```diff
@@ -75,6 +76,50 @@ type GoBackNConn struct {
+	// awaitedNACKSignal is used to signal that we have received the awaited
+	// NACK after resending the queue. Once this signal is received, we can
+	// proceed to send new packets.
+	awaitedNACKSignal chan struct{}
```
Ah good idea, definitely agree that it's way better to let this logic be contained within the queue struct instead! I've moved these variables + the resend handling logic to the `queue` struct intead!

ViktorTigerstrom commented on gbn/gbn_conn.go:
```diff
@@ -557,7 +763,7 @@ func (g *GoBackNConn) receivePacketsForever() error { // nolint:gocyclo
 				if lastNackSeq == g.recvSeq &&
 					time.Since(lastNackTime) < g.resendTimeout {
-
+					log.Tracef("%d recently sent NACK for %d", g.seqTest2, m.Seq)
```
Thanks, updated!

ViktorTigerstrom commented on gbn/gbn_conn.go:
```diff
@@ -378,20 +425,179 @@ func (g *GoBackNConn) sendPacket(ctx context.Context, msg Message) error {
+//
+// When either of the 2 conditions above are met, we will consider both parties
+// to be in sync, and we can proceed to send new packets.
+func (g *GoBackNConn) resendQueue() error {
```
Great idea, moved!

ViktorTigerstrom commented on gbn/queue.go:
```diff
@@ -119,13 +123,25 @@ func (q *queue) resend(cb func(packet *PacketData) error) error {
 
+// isPayloadPackets returns true if the packets for the passed packet indexes
+// isn't Ping packets.
+func (q *queue) isPayloadPackets(base, top uint8) bool {
```
Thanks, good idea!

ViktorTigerstrom commented on gbn/gbn_conn.go:
```diff
@@ -75,6 +76,50 @@ type GoBackNConn struct {
+	// completed through an awaitedNACKSignal, doesn't get marked as
+	// completed by a delayed proceedAfterTime callback from the previous
+	// resend catch up cycle.
+	catchUpId int64
```
See https://github.com/lightninglabs/lightning-node-connect/pull/87#discussion_r1391462799 (2)

ViktorTigerstrom commented on gbn/queue.go:
```diff
@@ -78,7 +143,66 @@ func (q *queue) addPacket(packet *PacketData) {
+//
+// When either of the 2 conditions above are met, we will consider both parties
+// to be in sync, and we can proceed to send new packets.
+func (q *queue) resend(resendTimeout time.Duration, quit chan struct{},
+	sendFunc func(packet *PacketData) error,
+) error {
```
The linter requires the signature to have this syntax to be accepted, but this collides with how I've learned that we deal with the return type when the function declaration is multiple lines.
Is this version ok, or should I move the return type up to the second line despite what the linter suggests?

ViktorTigerstrom replied:
For some reason that wasn't accepted by the linter, but this is now fixed by your suggestion of passing it through the config in the constructor instead :)! Thanks!

ViktorTigerstrom reviewed (commented):
Thanks a lot for the review @ellemouton!!

I've addressed your feedback, and I totally agree that it makes much more sense to let the queue struct handle the resend catch up logic instead, so I've moved it there!

Please also check my response here, and let me know what you'd like me to do with the `catchUpID`: https://github.com/lightninglabs/lightning-node-connect/pull/87#discussion_r1391462799

ViktorTigerstrom commented:
The itest CI that failed this round of the CI workflow is pre-existing since before this PR. It works sometimes, and fails sometimes.

ellemouton commented:
> The itest CI that failed this round of the CI workflow is pre-existing since before this PR. It works sometimes, and fails sometimes.

You are able to trigger a manual re-run right?

Also - can you make an issue for this flake so that some investigation can be done?

ViktorTigerstrom commented:
I've learned how to trigger a re-run of the workflow now :)! Also opened a new issue for the error. Thanks for the nudge :)!

ellemouton reviewed (commented):
check out this patch for how we can do a bit of cleanup: [gbn_cleanup.patch](https://github.com/lightninglabs/lightning-node-connect/files/13363308/gbn_cleanup.patch)

Left a few comments - Im mainly concerned with adding complexity to the system to solve this issue. I think we should maybe first focus on reducing the complexity and then tackle the issue at hand. 

So im thinking we should really try to divide up what exactly is being handled by GBN and what is being handled by `queue`. I think we can make use of a state-machine model too. Currently it seems like the complexity is coming from the fact that both the queue and the GBN can tricker timeouts & resends. I think this should only be handled by one entity. So for example, if we are in the "resend&sync" state, then nothing can trigger a new resent&sync state since we are already in that state. 

Let me know what you think.

ViktorTigerstrom commented on gbn/gbn_conn.go:
```diff
@@ -578,7 +591,10 @@ func (g *GoBackNConn) receivePacketsForever() error { // nolint:gocyclo
 		case *PacketACK:
-			gotValidACK := g.sendQueue.processACK(m.Seq)
+			gotValidACK := g.sendQueue.processACK(
+				m.Seq, g.resendTimeout,
```
So the problem with just setting the `resendTimeout` through the constructor, is that it the queue's `resendTimeout` then currently won't get set to the value set by the `WithTimeout` option.

Though I've fixed that issue in #88 now, by the addition of the `TimeoutManager`. If you agree, we'll let this PR still pass the timeout in the different function calls to the queue, and then remove that with #88 instead?

ViktorTigerstrom commented:
Getting a lint error similar to https://github.com/lightninglabs/lightning-node-connect/pull/87#discussion_r1391472194 now, but for the `newGoBackNConn` constructor instead. The function signature is correctly formatted though according to me. Is there anything I've missunderstood?

ViktorTigerstrom commented:
Rebased on master and addressed the feedback above!

@ellemouton, we discussed the offline that it could make sense to split up the first commit a bit, and add the patch you included previously as a first commit to this PR. Though after the rebase on master, much of that patch has now become outdated and been removed to a large extent. Including that as a separate commit would then basically only add a `Config` to the queue with the `s` value only. So I opted to not make that as separate commit as that seems extensive on the other end.

I'm not really sure how I should split the commit now. Feel free to suggest some way of splitting it in case you think it's still needed after the rebase :)!

guggero commented on gbn/queue.go:
```diff
@@ -112,28 +236,116 @@ func (q *queue) resend(cb func(packet *PacketData) error) error {
+//
+//nolint:cyclop
+func (q *queue) awaitCatchUp(resendTimeout time.Duration) {
+	ticker := time.NewTimer(resendTimeout * awaitingTimeoutMultiplier)
```
I think this can just be a `time.After()`, no need to start and then stop a timer.

guggero commented on gbn/queue.go:
```diff
@@ -112,28 +236,116 @@ func (q *queue) resend(cb func(packet *PacketData) error) error {
+			q.awaitingCatchUpMu.Unlock()
+
+			break catchupLoop
+		default:
```
If we have a default case in a `select`, it basically means this will spin in a loop very very quickly and eat up a lot of CPU. Since we have a timeout ticker I think we can just remove the `default` case here.

ViktorTigerstrom replied:
Ah thanks, you're right :)!

guggero commented on gbn/queue.go:
```diff
@@ -219,6 +468,45 @@ func (q *queue) processNACK(seq uint8) (bool, bool) {
+// proceedAfterTime will wait for the resendTimeout and then send an
+// awaitedACKSignal, if we're still awaiting the resend catch up.
+func (q *queue) proceedAfterTime(resendTimeout time.Duration) {
+	go func() {
```
Should we instead call `go proceedAfterTime()` to make it more obvious at the call site that we're spinning up a goroutine?

ViktorTigerstrom replied:
Good idea, thanks!

guggero commented on gbn/queue.go:
```diff
@@ -219,6 +468,45 @@ func (q *queue) processNACK(seq uint8) (bool, bool) {
+// awaitedACKSignal, if we're still awaiting the resend catch up.
+func (q *queue) proceedAfterTime(resendTimeout time.Duration) {
+	go func() {
+		ticker := time.NewTicker(resendTimeout)
```
Same here re using the tiny bit more lightweight `time.After()`, since we're never resetting the timer.

guggero commented on gbn/queue.go:
```diff
@@ -41,29 +65,71 @@ type queue struct {
+	awaitingCatchUp bool
+
+	// awaitingCatchUpMu must be held when accessing or mutating the values
+	// of, awaitedACK, catchUpID, awaitedNACK and awaitingCatchUp.
```
nit: catchUpID is no longer a struct member.

ellemouton commented:
> Including that as a separate commit would then basically only add a Config to the queue with the s value only

I still think it is worth it to do it in a separate commit since it changes quite a few lines and so makes the bug fix commit hard to read. See my suggested commit [here](https://github.com/ellemouton/lightning-node-connect/pull/13/commits/c0eea2532a510ba14577a61e1f8d613f32cce68f)

> even though it makes the flow logic a bit more involved.

Yeah - im worried about GBN become too complex if we just add onto it as is. IMO, we should first modularise & clean up the GBN before adding new logic to what is already an overcomplicated piece of code. I took a stab at this today [here](https://github.com/ellemouton/lightning-node-connect/tree/pktResend) if you guys want to see what im envisioning. 

Basically it splits the GBN into more modular pieces by creating:
1) a `sender` which is only responsible for sending `DATA` packets and handing `ACK`s and `NACK`s
2) a `receiver` which is only responsible for receiving `DATA` and sending `ACK`s and `NACK`s
3) a `distributer` which sends `DATA` packets to the `receiver` and `ACKs` and `NACK`s to the `sender`

the idea is to keep the logic exactly as it is today, just modularising it. I really think it would be worth it to do this first as it will make it easier to reason about any new additions. 

Let me know what you guys think (cc @guggero ) - the other option of course is that we add the fix first and then do the modularisation after that

guggero commented:
> I really think it would be worth it to do this first as it will make it easier to reason about any new additions.

I'm all for modularization! My gut feeling tells me that we should try to get all functionality that we know we want to have in, before doing a bigger refactor. That way we're more confident we can accommodate all functionality in the new, refactored model. Otherwise we have might have to change the new model you described above to integrate the bug fix in this PR and #88. Or are you saying your branch already contains both PRs, @ellemouton?

ellemouton commented:
cool - sounds good @guggero  👍

ellemouton commented:
is this ready for another round @ViktorTigerstrom ?

ViktorTigerstrom commented:
Thanks a lot for the review @guggero! I've addressed your feedback with the latest push. In regards to tests:
You're right that it's very hard to simulate a test that ensures that this doesn't go into to a forever ongoing resend loop with this change. I have an idea that I could potentially make an itest that simulates latency but that uses a local mailbox, and that the test should then verify that the exact amount of packets was sent that we expected after triggering a resend. Though it's not super straight forward to implement.
Another idea I've discussed with Elle before, is that this should be covered by itests timeouts if we can just set sane timeout values, as that should ensure that if we go into the resend loops the timeouts should potentially be triggered. It's not very easy to set sane timeouts , and currently they are clearly wrong as the itests CI keeps failing like 50% of the time.
For now though, I think that we need to keep this as a manual test and just check the logs that it doesn't go into an forever ongoing resend loop.

In regards to a unit test, I'll work on a queue unit test that triggers a resend just to see that we wait the expected time before / expected NACK before proceeding to send new packets. It should be noted though, that we have many itests that do trigger resends already, so it's covered that the added functionally at least doesn't cause unexpected panics etc.


And thanks a lot for the new review + suggested commit and cleanup @ellemouton!! I've added the first commit that you suggested, but left out the initial cleanup due to the conversation above :). Thanks a lot for creating the clean up though, I definitely agree that it modularizes things, and we should work on adding it after merging these new features :).

guggero commented:
Hey @ViktorTigerstrom. Yeah, makes sense for this to be quite tricky. And we do have pretty solid integration tests already indeed. So the proposed additional unit test sounds good to me.

ellemouton commented:
> They are currently still failing unfortunately, even after adding this fix.

And are we sure that the packet sending bug is not the cause in these failed ones - ie, can we see that cause is purely latency and not the cause of the bug?

ViktorTigerstrom commented on gbn/queue.go:
```diff
@@ -197,17 +409,40 @@ func (q *queue) processNACK(seq uint8) (bool, bool) {
+		// If we receive the awaited NACK, we shouldn't trigger a new
+		// resend, as we can now proceed to send new packets.
+		return false, false
+	}
```
Your right, with the current version it's covered as there's no way we'll complete the sync without the NACK seq being equal to the q.sequenceTop :)

ViktorTigerstrom commented on gbn/queue.go:
```diff
@@ -114,28 +236,100 @@ func (q *queue) resend() error {
+	q.caughtUpSignal <- struct{}{}
+}
+
 // processACK processes an incoming ACK of a given sequence number.
```
Updated with the latest push!

ViktorTigerstrom commented:
Thanks a lot for taking the time to drafting the refactor to break out the syncing logic into a separate struct @ellemouton! I've now added it to this PR, after adjusting it slightly to cover some edge cases :rocket:!

Let me know if you'd like me to remove the initial resending loop protection commits, or if you think it makes sense to keep this refactor as a final separate commit.

ViktorTigerstrom commented on gbn/queue.go:
```diff
@@ -114,28 +236,100 @@ func (q *queue) resend() error {
+	// packets. That could mean that the NACK that the resent packets will
+	// trigger, might be received after the timeout has passed. That would
+	// cause the resend loop to trigger once more.
+	q.awaitingCatchUpMu.Unlock()
```
Thanks, I've updated your syncer slightly to also be able to process the awaited ACKs + NACKs while we are actually resending the queue, and not block by holding the lock while actually resending the queue.
This new version can trigger the edge case that the NACK that the resent packets will trigger, might be received after the proceedAfterTime timeout has passed. But as this would only cause one extra resend loop in the rare case that this would happen, I think it's worth the tradeoff :)!

ViktorTigerstrom commented on gbn/syncer.go:
```diff
@@ -0,0 +1,291 @@
+		// As we can't be sure that waitForSync cancel listener was
+		// triggered before this one, we send over the cancel channel
+		// again, to make sure that both listeners are triggered.
+		c.reset()
```
We could potentially cover this instead by closing the `cancel` channel when executing `reset` function, and creating a new channel on every sync init (and also use a `sync.Once` to ensure we only cancel the channel once per sync). But I figured this version is probably a bit more clean.

ViktorTigerstrom replied:
> But then you would need to re-init the sync.Once each time too which kinda defeats the purpose right?

Yes exactly, which is why I opted for this solution :)!

> If you want to make sure that the reset only does something if we actually need it to,
then, in the reset method, you could check that we are in the "Resending" state, if we are not, it becomes a no-op. That way, the state machine becomes a bit more safe. we could add the same type of check anywhere else where we change the state

Hmmm I'm not sure how this would address that we wouldn't need to call the `c.reset()` both here and at the end of the `waitForSync` function as we can't be sure which of the select cases here + or in the `waitForSync` will catch the `<-c.cancel` case first? We need to ensure that both trigger when `c.cancel` channel get's sent over. Is this a case where you'd use a channel with a size of 2, or is there a better way to solve this?

ellemouton commented on gbn/queue.go:
```diff
@@ -112,36 +127,48 @@ func (q *queue) resend(cb func(packet *PacketData) error) error {
 	q.topMtx.RUnlock()
 
 	if base == top {
+		q.syncer.reset()
```
why do we need a `reset` here? 

Initially in my diff I had it here cause there was an `init` state above this. but if that is removed then this can be removed right?

ellemouton replied:
btw - the reason i had the other "init" state here was to remove the need for locking the awaitMu mutex before doing the seqBase and seqTop locks. 

if we still want there to be no other "resend" call during the time this is going, then we should add that back. 

but i think this is synchronous for now right?

ViktorTigerstrom replied:
> why do we need a reset here

Ah my bad, missed removing that. Thanks!

> btw - the reason i had the other "init" state here was to remove the need for locking the awaitMu mutex before doing the seqBase and seqTop locks.

Yeah, it's synchronous currently, so we will never call the `queue.resend` function twice at the same time while it's currently resending the function. 
Since we now process the awaited ACK+NACK while we are also resending packets, we now need to ensure that we set the `syncStateResending` while under the same lock that we also set the `awaitedACK` + `awaitedNACK` which is why I changed this. But like you say, if we add an extra init state to the syncer we could achieve that we wouldn't allow two concurrent calls of the `queue.resend` function. Opted to not add it now though to not cause extra confusion with unused functionality, butcan change it if you'd prefer :).

ellemouton commented on gbn/queue.go:
```diff
@@ -197,17 +417,40 @@ func (q *queue) processNACK(seq uint8) (bool, bool) {
 
 	// Is the NACKed sequence even in our queue?
 	if !containsSequence(q.sequenceBase, q.sequenceTop, seq) {
+		q.cfg.log.Tracef("NACK seq %d is not in the queue. Ignoring.",
```
it seems the "bumped" logic is changed here only to be changed again in a later commit...

If there is a bug in the original logic - perhaps better to fix that in it's own commit _before_ any additions. Cause these 2 diffs are a bit confusing to follow

ViktorTigerstrom replied:
Yeah good idea, cleaned this up with the latest push :)!

ellemouton reviewed (commented):
Leaving some comments :) 
I think your idea of first adding the syncer & then injecting the usage is a good one and will make reviewing easier since we then only need to review the logic once vs 2 versions of similar logic. 

One thing that concerns me is that it seems (from looking at the results of some of the latest GH action runs) that with this PR, the speed of the reconnection test has strictly been increased and so is resulting in more timeouts. 

```
before PR

    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:false,grpcClientConn:false) (9.82s)
    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:false,grpcClientConn:true) (14.81s)
    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:true,grpcClientConn:false) (12.78s)
    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:true,grpcClientConn:true) (16.00s)
    --- PASS: TestLightningNodeConnect/test_server_reconnect(stagingMailbox:false,grpcClientConn:false) (11.80s)
    --- PASS: TestLightningNodeConnect/test_server_reconnect(stagingMailbox:false,grpcClientConn:true) (19.80s)
    --- PASS: TestLightningNodeConnect/test_server_reconnect(stagingMailbox:true,grpcClientConn:false) (21.29s)
    --- PASS: TestLightningNodeConnect/test_server_reconnect(stagingMailbox:true,grpcClientConn:true) (19.86s)


    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:false,grpcClientConn:false) (9.82s)
    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:false,grpcClientConn:true) (13.81s)
    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:true,grpcClientConn:false) (10.95s)
    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:true,grpcClientConn:true) (14.66s)
    --- PASS: TestLightningNodeConnect/test_server_reconnect(stagingMailbox:false,grpcClientConn:false) (15.82s)
    --- PASS: TestLightningNodeConnect/test_server_reconnect(stagingMailbox:false,grpcClientConn:true) (19.80s)
    --- PASS: TestLightningNodeConnect/test_server_reconnect(stagingMailbox:true,grpcClientConn:false) (22.18s)
    --- PASS: TestLightningNodeConnect/test_server_reconnect(stagingMailbox:true,grpcClientConn:true) (18.52s)

    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:false,grpcClientConn:false) (9.82s)
    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:false,grpcClientConn:true) (13.84s)
    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:true,grpcClientConn:false) (12.71s)
    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:true,grpcClientConn:true) (15.70s)
    --- PASS: TestLightningNodeConnect/test_server_reconnect(stagingMailbox:false,grpcClientConn:false) (21.82s)
    --- FAIL: TestLightningNodeConnect/test_server_reconnect(stagingMailbox:false,grpcClientConn:true) (33.75s)


with PR

    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:false,grpcClientConn:false) (9.82s)
    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:false,grpcClientConn:true) (28.67s)
    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:true,grpcClientConn:false) (24.85s)
    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:true,grpcClientConn:true) (16.08s)
    --- PASS: TestLightningNodeConnect/test_server_reconnect(stagingMailbox:false,grpcClientConn:false) (27.85s)
    --- PASS: TestLightningNodeConnect/test_server_reconnect(stagingMailbox:false,grpcClientConn:true) (41.81s)
    --- PASS: TestLightningNodeConnect/test_server_reconnect(stagingMailbox:true,grpcClientConn:false) (23.67s)
    --- PASS: TestLightningNodeConnect/test_server_reconnect(stagingMailbox:true,grpcClientConn:true) (55.31s)

    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:false,grpcClientConn:false) (29.89s)
    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:false,grpcClientConn:true) (29.85s)
    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:true,grpcClientConn:false) (12.09s)
    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:true,grpcClientConn:true) (29.47s)
    --- PASS: TestLightningNodeConnect/test_server_reconnect(stagingMailbox:false,grpcClientConn:false) (38.83s)
    --- FAIL: TestLightningNodeConnect/test_server_reconnect(stagingMailbox:false,grpcClientConn:true) (35.50s)

    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:false,grpcClientConn:false) (14.83s)
    --- PASS: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:false,grpcClientConn:true) (33.87s)
    --- FAIL: TestLightningNodeConnect/test_client_reconnect(stagingMailbox:true,grpcClientConn:false) (33.06s)
    --- PASS: TestLightningNodeConnect/test_server_reconnect(stagingMailbox:false,grpcClientConn:false) (31.83s)
    --- PASS: TestLightningNodeConnect/test_server_reconnect(stagingMailbox:false,grpcClientConn:true) (39.80s)
    --- FAIL: TestLightningNodeConnect/test_server_reconnect(stagingMailbox:true,grpcClientConn:false) (32.93s)

```

Is this perhaps a necessary/expected increase due to the fact that we now back-off a bit?

ViktorTigerstrom commented on gbn/queue.go:
```diff
@@ -101,8 +118,6 @@ func (q *queue) resend(cb func(packet *PacketData) error) error {
 		return nil
 	}
 
-	q.lastResend = time.Now()
```
Ah good catch, my bad!

ViktorTigerstrom commented on gbn/queue.go:
```diff
@@ -86,13 +101,15 @@ func (q *queue) addPacket(packet *PacketData) {
+// two parties to be seen as synced; this may fail in which case the caller is
+// expected to call resend again.
+func (q *queue) resend() error {
+	if time.Since(q.lastResend) < q.cfg.timeout {
```
see https://github.com/lightninglabs/lightning-node-connect/pull/87#discussion_r1428654100

ViktorTigerstrom commented:
Thanks a lot for the review @ellemouton! I've updated the commit history quite extensively with the latest push, and addressed your feedback, so I hope this version looks better :)! :rocket:

> Is this perhaps a necessary/expected increase due to the fact that we now back-off a bit?

Yeah this was part of why I added the `IsPingPacket` logic earlier in the PR that we later remove, as these reconnect tests now often trigger a resend syncing with ping packet which will time out (i.e. await the `resendTimeout*3`). 
So it is expected, but a bit unfortunate. I'll experiment a bit with different timeouts later when addressing https://github.com/lightninglabs/lightning-node-connect/issues/89, or is the timeouts something you'd like me to update with this PR :)?

guggero reviewed (approved):
Code looks good to me, great work!
But it seems like the server or client disconnect integration tests fail pretty consistently. Do you think that's something that needs to be addressed with different timeouts (e.g. #88) or should they pass reliably for this PR too?

I'm happy to merge this as-is if one of the PRs down the road show them passing reliably.

lightninglabs-deploy commented:
@ellemouton: review reminder

lightninglabs-deploy commented:
@ellemouton: review reminder

ellemouton commented:
I think the tests are timing out since we now don’t just keep resending right? the test that times out is the reconnection test - meaning that packets are getting dropped. So before, we would resend more frequently & now we back-off quite a bit. 

So I think if we are confident that this PR is an improvement on reliability then we should just increase the itest timeout to whatever the average timeout required by that test is. Looks like 1 min should be fine right?

In order to ensure reliability, I suggest running a simulation overnight that keeps sending packets back and forth and occasionally does a restart of the aperture server - the test is successful if by the end of the test (after 12 hours or w/e) the client and server are still in sync & no resend loops are occurring

ViktorTigerstrom commented:
Ok so I've finally figured out what's going on in the itests! There were 2 core problems, 1 which was caused by a bug in this PR, and 1 which is pre-existing:

### Issue 1:
The bug which was caused by this PR mainly effected the `testClientReconnect` test and was caused by me missing that when calling `g.resendTicker.Reset(g.cfg.resendTimeout)` in the `resendQueue` func in `GoBackNConn.sendPacketsForever`, that doesn't also drain the ticker's channel if there's a tick that has occoured before reseting the timer.
That caused the `resendQueue` func to be called over and over again when the queue `syncer` times out, and the client is down and never sending any ACKS/NACKS back, as the ticker always ticks when the `syncer` times out.

I've addressed this with the latest push!

### Issue 2:
The problem is pre-existing, and is not made worse by this PR. This is effecting the `testServerReconnect` test, and is what happening for https://github.com/lightninglabs/lightning-node-connect/issues/89.

So the core issue is that when we stop the server here, the server sends a FIN packet, and then closes the connection:
https://github.com/lightninglabs/lightning-node-connect/blob/fa60171034b8f3f5da51611eddc9c74858a92051/itest/connection_test.go#L113

We then verify that client's connection status has change to `mailbox.ClientStatusSessionNotFound` on the line below.

But as there's a sliiiight delay on between when the FIN packet is sent, and when the connection is closed, 25-50% of the time when running the test, if the client receives the FIN packet from the server and calls `recvFromStream` in `GoBackNConn.receivePacketsForever` before the connection has been closed, the client will never move the connection status to `mailbox.ClientStatusSessionNotFound`. 

When client then processes the FIN packet, the client will then close down and never change the the connection status.

Therefore the predicate in the `testServerReconnect` test that verifies that the connection status changes to `mailbox.ClientStatusSessionNotFound` within the timeout will never be satisfied, no matter how long the timeout is set to.

For the cases when the test is run and instead actually succeeds, the connection is already closed when the connection `recvFromStream` call is made, and the status is changed as per the implementation here (as `c.transport.Recv(ctx)` will error):
https://github.com/lightninglabs/lightning-node-connect/blob/fa60171034b8f3f5da51611eddc9c74858a92051/mailbox/client_conn.go#L310

I'm not a 100% sure how we should handle this correctly, so would appreciate any feedback on how you think this should be handled best! One potential solution that I experimented with would be to trigger the mailbox `client_conn` to recheck the connection status once the client's `gbn_conn` processes a FIN packet in the receive function. That will change the client's connection status to change to `ClientStatusNotConnected` after the connection has been closed, so given that we change the test verification predicate to also accept that status, the test will pass in this case. Does that sound like a reasonable solution, or do you have a better suggestion?

----------------

Finally, it's expected that the itests should take a bit longer for the reconnection tests as we will now not resend as frequently and back-off quite a bit as @ellemouton points out above. However I actually don't think we need to increase any timeouts, and it's mainly the 2 issues I've detailed above that's been causing issues for the itests :rocket:!

> In order to ensure reliability, I suggest running a simulation overnight that keeps sending packets back and forth and occasionally does a restart of the aperture server - the test is successful if by the end of the test (after 12 hours or w/e) the client and server are still in sync & no resend loops are occurring

This is a great idea, I'll implement that simulation asap and let it run for a while as suggested :fire:!

ellemouton commented:
Really nice investigation @ViktorTigerstrom  🔥 🔥 

regarding how to solve for issue 2, could an `onFIN` callback work? see: 
[onFin.patch](https://github.com/lightninglabs/lightning-node-connect/files/13829682/onFin.patch)

ViktorTigerstrom commented on mailbox/client_conn.go:
```diff
@@ -157,20 +157,32 @@ func NewClientConn(ctx context.Context, sid [64]byte, serverHost string,
+		gbn.WithKeepalivePing(
+			gbnClientPingTimeout, gbnPongTimeout,
+		),
+		gbn.WithOnFIN(func() {
```
We could potentially add this to the `server_conn.go` version as well, and force the server to set the connection status to status `ServerStatusIdle` as well. But I haven't really seen any case where this has caused any issues on the server side yet, so opted to not add it. But Will add it to the server side as well if you think I should!.

ViktorTigerstrom reviewed (commented):
Thanks a lot for the patch @ellemouton! That worked really well :fire::rocket:!
I've added it on top of this PR, so itests are now finally consistently passing :tada:!

guggero reviewed (approved):
Awesome investigation on the itest flake hunt! Very nice to have things pass so consistently now.

Great work, LGTM :tada:

ellemouton reviewed (approved):
tACK, LGTM! Great investigations & great work!! 💥 🚀 

The majority of my comments are spelling/grammar nits

ellemouton commented on gbn/syncer.go:
```diff
@@ -0,0 +1,300 @@
+	syncStateResending
+)
+
+// syncer is used to ensure that both the sender and the receiver are in sync
```
I just did a basic local tests & things seem to work fine :)

ViktorTigerstrom commented on gbn/queue.go:
```diff
@@ -197,23 +203,29 @@ func (q *queue) processNACK(seq uint8) (bool, bool) {
-		return true, false
+		bumped = true
+
+		return false, bumped
```
So the reason I changed this, was because it seemed like it was quite confusing the second boolean actually ment in the function prior to this change. It had slipped through previously the function returned `false` here for the "bumped" return boolean, even though it actually did bump. So I just wanted to add this to ensure that there's less chance of that happening again with future changes. I'm totally fine with changing this as you've suggested though if you think that's unnecessary 🙂!

ellemouton replied:
yeah I think it is the responsibility of the function comment description to explain what each return value means.

ViktorTigerstrom replied:
Ok, makes sense! Updated :)

ViktorTigerstrom commented on gbn/syncer.go:
```diff
@@ -0,0 +1,300 @@
+// response to that NACK would cause the resend loop to continue indefinitely.
+//
+// When either of the 2 conditions above are met, we will consider both parties
+// to be in sync.
```
Thanks! Added a timeout clause

ViktorTigerstrom commented on gbn/syncer.go:
```diff
@@ -0,0 +1,300 @@
+	// and then proceed to send new packets, unless we receive the
+	// awaitedNACK during the wait time. If that happens, we will proceed
+	// send new packets as soon as we have processed the NACK.
+	awaitedACK uint8
```
Updated 😃

ViktorTigerstrom commented on gbn/syncer.go:
```diff
@@ -0,0 +1,300 @@
+
+	// awaitedACK defines the sequence number for the last packet in the
+	// resend queue. If we receive an ACK for this sequence number during
+	// waiting to sync, we wait for the duration of the resend timeout,
```
I updated this to "number while waiting **_to_** sync", as I was unsure if "for" is correct in this case. If it is correct to use "for", let me know and I'll update!

ViktorTigerstrom commented on gbn/gbn_conn.go:
```diff
@@ -374,7 +376,25 @@ func (g *GoBackNConn) sendPacket(ctx context.Context, msg Message) error {
 	// resendQueue re-sends the current contents of the queue.
 	resendQueue := func() error {
-		return g.sendQueue.resend()
+		err := g.sendQueue.resend()
```
Hmmm sorry I don't fully understand what this is suggesting that I should do? Would you like me to include a cleanup in the `resendQueue` func here if it the `err := g.sendQueue.resend()` yields an error?

ellemouton replied:
im just saying that it is strange to not check the error immediately after calling g.sendQueue.resend(). 

So it is weird to have this pattern:

```
err = somethingThatCouldError()

<other logic> 

if err != nil {
    <do thing with error> 
}
```

It's better to have the structure of:

```
err = somethingThatCouldError()
if err != nil {
    <do thing with error> 
}

<other logic> 
```

The difference between A & B here though is that the <other logic> always gets called in A whereas it wont necessarily get called in B. So im saying that if we go with format B and you want to make sure that the <other logic> gets called - just add that other logic here:

```
err := g.sendPacketsForever()
if err != nil {
        <other logic> 
}
```

since that will get called if we do get an error on `g.sendQueue.resend()`

does that make sense?

ViktorTigerstrom replied:
Ah, now I understand what you were referring to, thanks!

I changed this to now return the error immediately after `g.sendQueue.resend()` has been executed if it errors.

It wasn't intentional previously that rest of the code in `resendQueue` _MUST_ run even if `g.sendQueue.resend()` had errored, it was more a case of that it didn't matter wether it did or not and it then previously skipped adding an extra if case. But I understand now why this is bad code design, thanks :)!

ViktorTigerstrom commented:
Thanks a lot for the review @ellemouton!
Left some comments on mainly https://github.com/lightninglabs/lightning-node-connect/pull/87#discussion_r1442577617 & https://github.com/lightninglabs/lightning-node-connect/pull/87#discussion_r1442604224. Once that has been resolved, I think this should be good to go!
